### DISCUSSION TITLE
CityJSON import

### DISCUSSION THREAD

**Discussion by phtj** [cityjson/specs#93] [Show and tell]
Hi all,

Today, I created an importer for CityJSON, for [Mobius Modeller](https://mobius.design-automation.net/). It is great that the job can be done in a day from start to finish, thanks to the clear spec! (OK, instances and textures not done yet, but I am still happy with the days work.) You can take a look at this example model, it is the Den Haag model from your dataset. 

- [Open model](https://mobius-08.design-automation.net/editor?file=f619dc1c65c390b1d1dfe7975b17bdc5ab20bc8b58e53a981decf30fe25c1458&node=1&defaultViewer=geo) and click the 'play' button top left.

![image](https://user-images.githubusercontent.com/8304948/139849332-721bba65-fbe3-4d62-87df-fff72f4d4073.png)

Two questions.

1) For some reason, I cannot directly read the GeoJSON data from your URL. If I copy the data to GitHub, I can read it directly from the "raw" GitHub url. It would be nice if we could have a "raw" web link.

2) I am doing everything web based, and using porj4js to project between coord systems. However, i am wondering of there is a way (using javascript, without hardcoding it) to go from 
- `urn:ogc:def:crs:EPSG::7415` to 
- `"+proj=sterea +lat_0=52.15616055555555 +lon_0=5.38763888888889 +k=0.9999079 +x_0=155000 +y_0=463000 +ellps=bessel +towgs84=565.417,50.3319,465.552,-0.398957,0.343988,-1.8774,4.0725 +units=m +vunits=m +no_defs"`

This actually has nothing to do with CityJSON. But I thought someone might know a solution. At the moment the user has to specify this nasty string, which is obviously problematic. They need to open the CityJSON file, search for `referenceSystem`, search google for `EPSG::7415`, then in the EPSG website copy and paste the proj4js string. Hmm... there must be a better way. Of course I can do defs for lots of projections... But I am wondering, is there already something out there? Or a web service?

Patrick

**Comment by hugoledoux** [comment]
Nice to see our specs allowed you to do all this in 1 day 🙌🏼

**Q1**: you mean that https://3d.bk.tudelft.nl/opendata/cityjson/1.0/DenHaag_01.json cannot be read because of CORS I assume, right? Good point, I will dump all the example datasets to a github repo then

**Q2**: I don't really know. But I guess you saw this? https://github.com/proj4js/proj4js#named-projections

**Reply by phtj** [reply]
Thanks for the reply.

**Q1**: Yes, that is right - it is CORS. 

**Q2**: Yes, proj4js allows you to add named projections. But you have to add the projections yourself. I.e. I would have to add every possible EPSG code, which would not be so great. I think there are around 4k to 5k.

I have been doing a bit more research into this. I found out that I can do an HTTP request to epsg.io like this, [https://epsg.io/7415.js](https://epsg.io/7415.js). So i can take the CRS string in the CityJSON file, extract the EPSG number out of it, and then do a HTTP request to epsg.io, and it will return text that contains the prjo4js string. So then I can use that to do my projection. This seems to work fine. It means my importer will only work with EPSG CRS. I guess that should be ok.

Here is [the updated model](https://mobius-08.design-automation.net/editor?file=380c3d0db84dafffd63cf2a47984eda7b79d85ab20088701494c7cb8191b39ca&amp;node=1&amp;defaultViewer=geo), now there is no need to specify the projection.

**Comment by hugoledoux** [comment]
I dumped the files in that repo: https://github.com/cityjson/example-datasets

I had to learn how LFS works because some are >100MB so thank you for forcing me to do this... The v1.1 files are there too btw, because next week or so v1.1 is launched (files almost the same, no worries. Draft of website: https://homepage.tudelft.nl/23t4p/cj11website/tutorials/upgrade11/)

Good to know about the epsg.io trick!

**Reply by phtj** [reply]
Thanks, it works.

Here are some of the examples of CityJSON 1.1:

- [Rotterdam](https://mobius-08.design-automation.net/editor?file=0aa4e8a24d1631d0b867fe972f80485674ec7db2a7545d1dc741f384206403f3&amp;node=1&amp;defaultViewer=geo)
- [Den Haag](https://mobius-08.design-automation.net/editor?file=369bfdb8eeea9f10241ec05ca63179581cf12030c6ee2f254d7289f045b8e902&amp;node=1&amp;defaultViewer=geo)
- [Vienna](https://mobius-08.design-automation.net/editor?file=5f52cc26b71deefe4d88ec8b5986ef1b03bd0129dfbcf773387bd9e61ae8ef36&amp;node=1&amp;defaultViewer=cad)
- [Montreal](https://mobius-08.design-automation.net/editor?file=013c0468d5c3e095039efd5f5b7376edba7e63470cf9074e5bee6368e7e33caf&amp;node=1&amp;defaultViewer=cad)

**Reply by phtj** [reply]
Also, these seems to be a discrepancy here: https://github.com/cityjson/specs/blob/master/changelog.md
`change CRS to 2 simple values: authority:code => EPSG:28992`

But in the Den Haag file:
`"referenceSystem": "https://www.opengis.net/def/crs/EPSG/0/7415"`

And in Rotterdam file:
`"referenceSystem":"https://www.opengis.net/def/crs/EPSG/0/28992"`

**Reply by hugoledoux** [reply]
Thanks, glad it,s working.

And yeah we moved to the "modern" OGC way finally (URNs were the old way apparently), which is with a URL

See this: https://homepage.tudelft.nl/23t4p/cj11website/dev/upgrade11/#4-metadatareferencesystem-is-now-using-ogc-urls

**Comment by hugoledoux** [comment]
@phtj bad news, I had to delete https://github.com/cityjson/example-datasets since it was taking more space than github allows for free and they sent me a bill! LFS is not free, fair enough...

so I dumped everything back to our server, which doesn't allow CORS: https://3d.bk.tudelft.nl/opendata/cityjson/

will try to find a (free) solution...
